src/workerd/api/pyodide/pyodide.h

```diff
@@ -1,45 +1,185 @@
+      return 0;
     }
     int toCopy = buf.size();
     if (tarSize - offset < toCopy) {
       toCopy = tarSize - offset;
     }
     memcpy(buf.begin(), &((*PYODIDE_PACKAGES_TAR)[0]) + offset, toCopy);
+    return toCopy;
   }
 
   JSG_RESOURCE_TYPE(PackagesTarReader) {
     JSG_METHOD(read);
   }
 };
 
+// A function to read a segment of the tar file into a buffer
+// Set up this way to avoid copying files that aren't accessed.
+class PyodideMetadataReader : public jsg::Object {
+private:
+  kj::String mainModule;
+  kj::Array<kj::String> names;
+  kj::Array<kj::Array<kj::byte>> contents;
+  kj::Array<kj::String> requirements;
+  bool isWorkerdFlag;
+
+public:
+  PyodideMetadataReader(kj::String mainModule, kj::Array<kj::String> names,
+                        kj::Array<kj::Array<kj::byte>> contents, kj::Array<kj::String> requirements,
+                        bool isWorkerd) {
+    this->mainModule = kj::mv(mainModule);
+    this->names = kj::mv(names);
+    this->contents = kj::mv(contents);
+    this->requirements = kj::mv(requirements);
+    this->isWorkerdFlag = isWorkerd;
+  }
+
+  bool isWorkerd() {
+    return this->isWorkerdFlag;
+  }
+
+  kj::String getMainModule() {
+    return kj::str(this->mainModule);
+  }
+
+  kj::Array<jsg::JsRef<jsg::JsString>> getNames(jsg::Lock& js) {
+    auto builder = kj::heapArrayBuilder<jsg::JsRef<jsg::JsString>>(this->names.size());
+    for (auto i : kj::zeroTo(builder.capacity())) {
+      builder.add(js, js.str(this->names[i]));
+    }
+    return builder.finish();
+  }
+
+  kj::Array<jsg::JsRef<jsg::JsString>> getRequirements(jsg::Lock& js) {
+    auto builder = kj::heapArrayBuilder<jsg::JsRef<jsg::JsString>>(this->requirements.size());
+    for (auto i : kj::zeroTo(builder.capacity())) {
+      builder.add(js, js.str(this->requirements[i]));
+    }
+    return builder.finish();
+  }
+
+  kj::Array<int> getSizes(jsg::Lock& js) {
+    auto builder = kj::heapArrayBuilder<int>(this->names.size());
+    for (auto i : kj::zeroTo(builder.capacity())) {
+      builder.add(this->contents[i].size());
+    }
+    return builder.finish();
+  }
+
+  int read(jsg::Lock& js, int index, int offset, kj::Array<kj::byte> buf) {
+    if (index >= contents.size() || index < 0) {
+      return 0;
+    }
+    auto& data = contents[index];
+    int dataSize = data.size();
+    if (offset >= dataSize || offset < 0) {
+      return 0;
+    }
+    int toCopy = buf.size();
+    if (dataSize - offset < toCopy) {
+      toCopy = dataSize - offset;
+    }
+    memcpy(buf.begin(), &data[0] + offset, toCopy);
+    return toCopy;
+  }
+
+  JSG_RESOURCE_TYPE(PyodideMetadataReader) {
+    JSG_METHOD(isWorkerd);
+    JSG_METHOD(getMainModule);
+    JSG_METHOD(getRequirements);
+    JSG_METHOD(getNames);
+    JSG_METHOD(getSizes);
+    JSG_METHOD(read);
+  }
+};
+
+using Worker = server::config::Worker;
+
+inline jsg::Ref<PyodideMetadataReader> makePyodideMetadataReader(Worker::Reader conf) {
+  auto modules = conf.getModules();
```

Comment:
This method is rather too large to be inlined. Would much rather this be moved out into a `pyodide.c++`

Reply:
Okay I did this.